scene/main/node.cpp

```diff
@@ -1731,10 +1750,23 @@ Node *Node::get_parent() const {
 	return data.parent;
 }
 
-Node *Node::find_parent(const String &p_pattern) const {
+// Finds the first parent node (in tree order) whose name matches the given
+// pattern, class name, or both (either pattern or type can be left empty).
+Node *Node::find_parent(const String &p_pattern, const String &p_type) const {
 	ERR_THREAD_GUARD_V(nullptr);
+	ERR_FAIL_COND_V(p_pattern.is_empty() && p_type.is_empty(), nullptr);
 	Node *p = data.parent;
 	while (p) {
+		if (p_pattern.is_empty() || p->data.name.operator String().match(p_pattern)) {
```

Comment:
Could `p->data.name.operator String()` be written as `String(p->data.name)`?

Also, the same match is computed further below in line 1770, might be worth storing. Although it admittedly gets a bit complicated since the computation is only necessary in the non-empty case...

The same applies above, in line 1682.

Reply:
I can't think of a way to cache it in a good way.

* If it's cached it will run once per iteration.
* If it's not cached it will run 0-2 times per iteration.

Reply:
The easiest would be a closure (lazy evaluation), but I'm not sure if that fits Godot code style. 

Maybe one of the maintainers can state how these things are usually addressed.